### PR TITLE
Redesign bookmarks with dual modes and server-side engagement tracking

### PR DESCRIPTION
**Pull request by WilliamAGH**
This pull request adds new and enhanced test coverage for the bookmarks and engagement API endpoints, with a focus on improving reliability, mocking, and semantic tag filtering. The changes introduce comprehensive tests for different bookmark feed modes, tag filtering behaviors (including semantic and related tags), and the engagement event API. Additionally, test setup is streamlined by centralizing mock data and request creation, and database dependencies are consistently mocked across test files.

**New and enhanced API test coverage**

* Added a new test suite for the engagement API (`__tests__/api/engagement.test.ts`), covering validation, DNT handling, and event batch processing, with fully mocked database interactions.
* Introduced a test suite for the bookmarks feed endpoint (`__tests__/api/bookmarks-feed.test.ts`), validating both chronological ("latest") and scored ("discover") feed modes, category diversity, and fallback logic when discovery ranking fails.

**Tag filtering and semantic relevance improvements**

* Refactored tag filtering tests in `__tests__/api/bookmarks/tag-filtering.test.ts` to use helper functions for index data and request creation, and added coverage for semantic tag expansion (appending related bookmarks after exact tag matches). [[1]](diffhunk://#diff-82c09e8e629e28bd82459ee615b3a92cadb57114da8533c8f5038aa891a9fd46R65-R85) [[2]](diffhunk://#diff-82c09e8e629e28bd82459ee615b3a92cadb57114da8533c8f5038aa891a9fd46L93-R181)
* Enhanced tag filtering tests to check for multi-word, URL-encoded, and case-insensitive tags, as well as edge cases like non-existent tags and pagination, using the new helpers for consistency and clarity. [[1]](diffhunk://#diff-82c09e8e629e28bd82459ee615b3a92cadb57114da8533c8f5038aa891a9fd46L150-R193) [[2]](diffhunk://#diff-82c09e8e629e28bd82459ee615b3a92cadb57114da8533c8f5038aa891a9fd46L178-R205) [[3]](diffhunk://#diff-82c09e8e629e28bd82459ee615b3a92cadb57114da8533c8f5038aa891a9fd46L206-R217) [[4]](diffhunk://#diff-82c09e8e629e28bd82459ee615b3a92cadb57114da8533c8f5038aa891a9fd46L249-R246)

**Test setup and mocking consistency**

* Standardized mocking of database connections and query dependencies across test files to avoid reliance on a real database and improve isolation. [[1]](diffhunk://#diff-850bcdb1ff0828bd6f551b0cc091a50210eb83d03b676ce57a6e64a977a5ef47R12-R21) [[2]](diffhunk://#diff-82c09e8e629e28bd82459ee615b3a92cadb57114da8533c8f5038aa891a9fd46L6-R34)
* Updated pagination limit tests to explicitly specify feed mode and use consistent request construction, ensuring accurate coverage of large dataset handling.

These changes significantly improve the reliability and maintainability of API tests, while expanding coverage to include semantic relevance, error handling, and real-world request scenarios.

---

> [!NOTE]
> **High Risk**
> Introduces a new engagement ingestion endpoint and changes the default `/api/bookmarks` behavior (new discover mode, caching headers, semantic expansion), which can impact traffic, DB load, and ranking correctness. Also adjusts SEO/sitemap behavior by retiring root bookmark pagination, requiring careful validation of redirects and indexing.
> 
> **Overview**
> Adds **dual-mode bookmarks feeds**: `/api/bookmarks` now supports `feed=discover|latest` (defaulting to *discover*), including an optional grouped discover view, engagement-weighted ranking via `getDiscoveryRankedBookmarks`, and explicit fallback to latest ordering when ranking data is unavailable.
> 
> Introduces **server-side engagement tracking** via `POST /api/engagement` with Zod validation, DNT short-circuiting, visitor hashing, and per-visitor rate limiting; the bookmarks API also gains *semantic tag expansion* (appending embedding-similar bookmarks after exact tag matches on page 1).
> 
> Updates SEO and routing around the new model by **retiring root `/bookmarks/page/[n]` sitemap/metadata coverage** and adding a legacy redirect expectation, alongside broad new/updated tests covering feed modes, tag filtering, engagement ingestion, discovery grouping/scoring utilities, and related helper behavior.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit edee8b1599c1765b818961c481d05f0a274d6a2d. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

<h3>Greptile Summary</h3>

This PR implements a comprehensive bookmarks discovery system with dual-mode feeds and server-side engagement tracking. The implementation is well-architected with proper separation of concerns, extensive test coverage, and graceful degradation.

## Key Changes

**Engagement Tracking Infrastructure**
- Adds `content_engagement` table tracking impression, click, dwell, and external_click events
- Server-side API (`/api/engagement`) with DNT support, rate limiting (100 events/60s), and visitor hash-based privacy
- Client-side hook batching impressions every 30s using `sendBeacon` with fetch fallback
- Comprehensive test coverage for validation, DNT compliance, and batching behavior

**Dual-Mode Bookmark Feed**
- "Latest" mode: chronological ordering (existing behavior)
- "Discover" mode: engagement-weighted scoring (82% recency + 18% engagement signals)
- Grouped view: topic sections with embedding-based cohesion scoring
- Semantic tag expansion: appends related bookmarks when exact tag matches are insufficient

**Discovery Scoring System**
- Novelty boost for bookmarks with <5 impressions
- Recency decay using exponential function (30-day half-life for engagement, 21-day for base recency)
- Engagement signal combines CTR (45%), dwell time (35%), and external clicks (20%)
- Deterministic tie-breaking using `dateBookmarked` comparison
- Multi-level fallbacks when engagement data or embeddings unavailable

**Tag Taxonomy Migration**
- Migrates `bookmark_categories` to new `bookmarks_tags` + `bookmarks_tags_links` tables
- Preserves existing data and establishes canonical tag relationships
- Supports primary tags and alias resolution

**Test Coverage**
- New test suites for engagement API, bookmark feed modes, and engagement tracker hook
- Refactored tag filtering tests with semantic expansion coverage
- Unit tests for discovery scoring, grouping, and embedding similarity
- Consistent database mocking patterns across all test files

## Architecture Strengths

- Graceful degradation at every level (engagement unavailable → recency-only, discovery fails → latest mode)
- Privacy-conscious design (DNT respected, visitor hashing, no PII stored)
- Performance optimizations (batching, indexes, caching headers)
- Separation of concerns (scoring, grouping, serialization in separate modules)
- Comprehensive error handling with descriptive degradation reasons

<h3>Confidence Score: 4/5</h3>

- Safe to merge with thorough testing of discovery feed and engagement tracking in staging
- Score reflects extensive test coverage, graceful degradation, and well-structured code, but this is a significant architectural change introducing new database tables, scoring algorithms, and user-facing features that should be validated in staging before production
- Pay close attention to database migrations (0019, 0021) in production deployment and monitor engagement API rate limiting effectiveness

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/app/api/engagement/route.ts | New engagement tracking API with DNT support, rate limiting (100 events/60s), and visitor hash-based tracking |
| src/hooks/use-engagement-tracker.ts | Client-side engagement hook batching impressions (30s intervals) and respecting DNT with beacon fallback |
| drizzle/0019_content-engagement.sql | Creates `content_engagement` table with proper indexes for content lookup, scoring queries, and visitor analysis |
| drizzle/0021_bookmark-tags-taxonomy.sql | Migrates `bookmark_categories` to new tag taxonomy with alias support and data preservation |
| src/app/api/bookmarks/route.ts | Dual-mode feed (latest/discover) with semantic tag expansion, grouped view support, and graceful degradation |
| src/lib/db/queries/discovery-scores.ts | Discovery ranking using recency (82%) + engagement (18%) with novelty boost and deterministic tie-breaking |
| src/lib/db/queries/discovery-grouped.ts | Topic sections with embedding-based cohesion scoring, pagination, and multi-level fallbacks |
| __tests__/api/engagement.test.ts | Tests validation, DNT handling, rate limiting, and batch event processing with mocked database |
| __tests__/api/bookmarks-feed.test.ts | Tests latest vs discover modes and fallback behavior when discovery ranking fails |
| __tests__/hooks/use-engagement-tracker.test.ts | Tests DNT compliance, impression batching, dwell timing, and cleanup behavior |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant User
    participant Client as useEngagementTracker
    participant API as /api/engagement
    participant DB as content_engagement
    participant Discovery as discovery-scores

    User->>Client: View bookmark
    Client->>Client: trackImpression()
    Note over Client: Batch in memory (30s interval)
    
    User->>Client: Click external link
    Client->>API: POST batch (sendBeacon)
    API->>API: Check DNT header
    alt DNT enabled
        API-->>Client: 204 (no tracking)
    else DNT disabled
        API->>API: buildVisitorHash(IP + UA)
        API->>DB: Check rate limit
        alt Rate limit exceeded
            API-->>Client: 429 (rate limited)
        else Within limit
            API->>DB: INSERT engagement events
            API-->>Client: 204 (success)
        end
    end

    Note over Discovery: Periodic scoring refresh
    Discovery->>DB: Query 90-day engagement
    Discovery->>Discovery: computeEngagementSignal()
    Discovery->>Discovery: computeDiscoveryScore()<br/>(82% recency + 18% engagement)
    
    User->>API: GET /api/bookmarks?feed=discover
    API->>Discovery: getDiscoveryRankedBookmarks()
    Discovery-->>API: Scored bookmarks
    API-->>User: Ranked feed with degradation status
```

<sub>Last reviewed commit: 9c498f9</sub>

### CODE REVIEW

File: __tests__/hooks/use-engagement-tracker.test.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+    expect(sendBeaconMock).toHaveBeenCalledTimes(1);
+    expect(sendBeaconMock).toHaveBeenCalledWith(
+      "/api/engagement",
+      expect.stringContaining('"eventType":"impression"'),
+    );
+
+    const payload = JSON.parse(sendBeaconMock.mock.calls[0]?.[1] as string) as {
+      events: Array<{ contentId: string; eventType: string }>;
+    };
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply __tests__/hooks/use-engagement-tracker.test.ts]: # Suggestion Implemented ✅
## Changes Made
- Replaced unchecked JSON casts with runtime validation via `engagementBatchSchema.parse(...)` in `__tests__/hooks/use-engagement-tracker.test.ts`.
## Validation
- `bun run validate` passed with 0 errors and 0 warnings.

File: src/components/features/bookmarks/category-ribbon.client.tsx [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    return [];
+  }
+
+  const candidate = payload as { categories?: unknown };
+  if (!Array.isArray(candidate.categories)) {
+    return [];
+  }
+
+  return candidate.categories.filter(isBookmarkCategorySummary);
+}
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/components/features/bookmarks/category-ribbon.client.tsx]: # Suggestion Analysis 🔍
## Decision: Not Implemented
## Reasoning
This thread targets `src/components/features/bookmarks/category-ribbon.client.tsx`, which is no longer present in the current PR head. The referenced code path is obsolete in the current diff.

File: src/lib/db/queries/content-hydration.ts [hunk 3]
```diff
@@ -85,28 +85,39 @@ async function hydrateBookmarks(entries: HydrationEntry[]): Promise<RelatedConte
...
-        }) ?? undefined,
-      ),
-    } satisfies RelatedContentMetadata,
-  }));
+  return rows.map((r) => {
+    // Related-content cards should only trust persisted CDN OG URLs.
+    // External OG URLs are intentionally excluded to avoid nondeterministic/hallucinated previews.
+    // Asset IDs in bookmark.content remain the canonical fallback chain.
+    const trustedOgImage = r.ogImage && isOurCdnUrl(r.ogImage, cdnConfig) ? r.ogImage : undefined;
+
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/lib/db/queries/content-hydration.ts]: # Suggestion Implemented ✅
## Changes Made
- Added explicit warning logs when non-CDN bookmark `ogImage` values are dropped.
- Added explicit warning when a hydration score is missing before defaulting to `0`.
## Validation
- `bun run validate` passed with 0 errors and 0 warnings.

File: src/lib/db/schema/bookmark-taxonomy.ts [hunk 4]
```diff
@@ -46,6 +46,41 @@ export const bookmarkTagLinks = pgTable(
...
+  },
+  (table) => [
+    primaryKey({
+      columns: [table.sourceTagSlug, table.targetTagSlug],
+      name: "bookmarks_tags_links_pk",
+    }),
+    index("idx_bookmarks_tags_links_source").on(table.sourceTagSlug),
+    index("idx_bookmarks_tags_links_target").on(table.targetTagSlug),
+  ],
+);
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/lib/db/schema/bookmark-taxonomy.ts]: # Suggestion Implemented ✅
## Changes Made
- Added unique index declaration for alias-source uniqueness in `bookmarkTagAliasLinks` (`idx_bookmarks_tags_links_source_alias`) to align schema-level constraints with canonical alias mapping expectations.
## Validation
- `bun run validate` passed with 0 errors and 0 warnings.

File: docs/features/bookmarks.md [hunk 5]
```diff
@@ -92,15 +115,54 @@ DEPLOYMENT_ENV=production NODE_ENV=production node scripts/backfill-bookmark-emb
...
+
+```
+Client events → POST /api/engagement → content_engagement table
+                                              ↓
+                        getDiscoveryRankedBookmarks() joins bookmarks + engagement aggregation
+                                              ↓
+                        applyCategoryDiversity() prevents 3+ consecutive same-category items
+                                              ↓
+                        Client SWR receives ranked bookmarks
+```
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply docs/features/bookmarks.md]: # Suggestion Implemented ✅
## Changes Made
- Added language specifier (`text`) to the fenced data-flow block in `docs/features/bookmarks.md`.
## Validation
- `bun run validate` passed with 0 errors and 0 warnings.

File: src/components/features/bookmarks/bookmark-card.client.tsx [hunk 6]
```diff
@@ -114,13 +71,110 @@ export function BookmarkCardClient(props: BookmarkCardClientProps): JSX.Element
...
+            {formattedBookmarkDate && (
+              <span
+                className="inline-flex items-center gap-1 rounded-full border border-gray-200 dark:border-gray-700 px-2 py-0.5 text-[11px]"
+                title={hoverDateLabel}
+                aria-label={hoverDateLabel}
+              >
+                <Calendar className="w-3 h-3" />
+                {compactBookmarkDate}
+              </span>
+            )}
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/components/features/bookmarks/bookmark-card.client.tsx]: # Suggestion Implemented ✅
## Changes Made
- Replaced the non-semantic date `<span>` with `<time dateTime={dateBookmarked}>` in `bookmark-card.client.tsx`, preserving tooltip text.
## Validation
- `bun run validate` passed with 0 errors and 0 warnings.

File: __tests__/api/bookmarks-feed.test.ts [hunk 7]
```diff
@@ -0,0 +1,146 @@
+import { GET } from "@/app/api/bookmarks/route";
+import { getBookmarksIndex, getBookmarksPage } from "@/lib/bookmarks/service.server";
+import { getDiscoveryRankedBookmarks } from "@/lib/db/queries/discovery-scores";
+import { loadSlugMapping } from "@/lib/bookmarks/slug-manager";
+import type { BookmarkSlugMapping, UnifiedBookmark } from "@/types";
+
+vi.mock("@/lib/bookmarks/service.server");
+vi.mock("@/lib/bookmarks/slug-manager");
+vi.mock("@/lib/db/queries/discovery-scores");
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply __tests__/api/bookmarks-feed.test.ts]: # Suggestion Implemented ✅
## Changes Made
- Added missing mocks in `__tests__/api/bookmarks-feed.test.ts` for transitive route dependencies: `discovery-grouped` and `embedding-similarity`.
## Validation
- `bun run validate` passed with 0 errors and 0 warnings.

File: src/components/ui/button.tsx [hunk 8]
```diff
@@ -0,0 +1,47 @@
+import * as React from "react";
+import { Slot } from "@radix-ui/react-slot";
+import { cva, type VariantProps } from "class-variance-authority";
+
+import { cn } from "@/lib/utils";
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/components/ui/button.tsx]: # Suggestion Analysis 🔍
## Decision: Not Implemented
## Reasoning
Per Next.js 16 docs (`use client` boundary semantics): once a parent entry file is marked `"use client"`, imported child components are included in the client bundle; they do not each require their own directive unless they are direct server-imported client entrypoints. `Button` is used as an imported child from client boundaries.

File: __tests__/app/bookmarks/page/page-number.test.tsx [hunk 9]
```diff
@@ -16,69 +15,130 @@ vi.mock("@/lib/bookmarks/bookmarks-data-access.server", () => ({
...
     await act(async () => {
-      const rendered = render(PageComponent as React.ReactElement);
+      const rendered = render(RootComponent as React.ReactElement);
       container = rendered.container;
     });
 
-    // Verify the component rendered
     expect(container).toBeTruthy();
+    expect(container?.querySelector('[data-testid="discover-feed"]')).not.toBeNull();
+    expect(mockGetDiscoveryGroupedBookmarks).toHaveBeenCalledTimes(1);
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply __tests__/app/bookmarks/page/page-number.test.tsx]: # Suggestion Analysis 🔍
## Decision: Not Implemented
## Reasoning
The retained assertions verify branch-specific data-fetch behavior for server-page routing decisions (`discover` vs `latest`) in a mocked-boundary test harness. Combined with DOM assertions, they guard correctness of route-mode behavior under this unit setup.

File: src/app/bookmarks/tags/[...slug]/page.tsx [hunk 10]
```diff
@@ -43,6 +44,39 @@ function parsePageParam(page: string | undefined, pageNumberStr: string | undefi
   return Number.isNaN(parsed) || parsed < 1 ? 1 : parsed;
 }
 
+function decodeTagSlug(rawTagSlug: string): string {
+  try {
+    return decodeURIComponent(rawTagSlug);
+  } catch {
+    return rawTagSlug;
+  }
+}
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/app/bookmarks/tags/[...slug]/page.tsx]: # Suggestion Implemented ✅
## Changes Made
- Added explicit warning log in `decodeTagSlug` catch path with slug + error context before fallback return.
## Validation
- `bun run validate` passed with 0 errors and 0 warnings.
